operator/src/lib.rs

```diff
@@ -0,0 +1,402 @@
+    pub fn get_full_pod_node_map(&self) -> Vec<(Vec<Node>, LabelOptionalValueMap)> {
+        let mut eligible_nodes_map = vec![];
+
+        for node_type in NiFiNodeType::iter() {
+            if let Some(eligible_nodes_for_role) = self.eligible_nodes.get(&node_type) {
+                for (group_name, eligible_nodes) in eligible_nodes_for_role {
+                    // Create labels to identify eligible nodes
+                    trace!(
+                        "Adding [{}] nodes to eligible node list for role [{}] and group [{}].",
+                        eligible_nodes.len(),
+                        node_type,
+                        group_name
+                    );
+                    eligible_nodes_map.push((
+                        eligible_nodes.clone(),
+                        get_node_and_group_labels(group_name, &node_type),
+                    ))
+                }
+            }
+        }
+        eligible_nodes_map
+    }
+
+    pub fn get_deletion_labels(&self) -> BTreeMap<String, Option<Vec<String>>> {
+        let roles = NiFiNodeType::iter()
+            .map(|role| role.to_string())
+            .collect::<Vec<_>>();
+        let mut mandatory_labels = BTreeMap::new();
+
+        mandatory_labels.insert(String::from(APP_COMPONENT_LABEL), Some(roles));
+        mandatory_labels.insert(String::from(APP_INSTANCE_LABEL), None);
+        mandatory_labels
+    }
+
+    async fn create_config_map(&self, name: &str, config: &NiFiConfig) -> Result<(), Error> {
+        match self
+            .context
+            .client
+            .get::<ConfigMap>(name, Some(&"default".to_string()))
+            .await
+        {
+            // TODO: check and compare content here
+            Ok(_) => {
+                debug!("ConfigMap [{}] already exists, skipping creation!", name);
+                return Ok(());
+            }
+            Err(e) => {
+                // TODO: This is shit, but works for now. If there is an actual error in comms with
+                //   K8S, it will most probably also occur further down and be properly handled
+                debug!("Error getting ConfigMap [{}]: [{:?}]", name, e);
+            }
+        }
+        let config = create_config_file(config);
+
+        let mut data = BTreeMap::new();
+        data.insert("nifi.properties".to_string(), config);
+
+        // And now create the actual ConfigMap
+        let cm =
+            stackable_operator::config_map::create_config_map(&self.context.resource, &name, data)?;
+        self.context.client.create(&cm).await?;
+        Ok(())
+    }
+
+    async fn create_missing_pods(&mut self) -> NiFiReconcileResult {
+        // The iteration happens in two stages here, to accommodate the way our operators think
+        // about nodes and roles.
+        // The hierarchy is:
+        // - Roles (for example Datanode, Namenode, NiFi Server)
+        //   - Node groups for this role (user defined)
+        //      - Individual nodes
```

Comment:
What's meant by a "node" here?

Reply:
In the selector we have:
        instances: 1
        instancesPerNode: 1
which is currently not used. I remove it for now.